### ISSUE TITLE
Blazor parameters should be case-sensitive

### ISSUE THREAD

**Issue by stsrki** [dotnet/aspnetcore#32252]
### Describe the bug

I'm trying to add some advanced _utility_ parameters to my component library and one of those utilities is used for component sizing. In this case, I have a `Width` parameter. The example of my component usage as follows:

### This Works!

```cshtml
<Button Width="Width.Max">Test</Button>
```

My custom utility works as expected but while testing I came to a strange bug. If I want to use a _regular_ or native `width` attribute for some reason it will try to convert it into my own custom object.

### This does not work!

```cshtml
<Button width="100">Test</Button>
```

and would give me an error (seen below).

### Exceptions

```
crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Unable to set property 'width' on object of type 'Blazorise.Bootstrap.Button'. The error was: Specified cast is not valid.
System.InvalidOperationException: Unable to set property 'width' on object of type 'Blazorise.Bootstrap.Button'. The error was: Specified cast is not valid.
 ---> System.InvalidCastException: Specified cast is not valid.
   at Microsoft.AspNetCore.Components.Reflection.PropertySetter.CallPropertySetter[Button,IFluentSizing](Action`2 setter, Object target, Object value)
   at Microsoft.AspNetCore.Components.Reflection.PropertySetter.SetValue(Object target, Object value)
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.<SetProperties>g__SetProperty|2_0(Object target, PropertySetter writer, String parameterName, Object value)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.<SetProperties>g__SetProperty|2_0(Object target, PropertySetter writer, String parameterName, Object value)
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.SetProperties(ParameterView& parameters, Object target)
   at Microsoft.AspNetCore.Components.ParameterView.SetParameterProperties(Object target)
   at Microsoft.AspNetCore.Components.ComponentBase.SetParametersAsync(ParameterView parameters)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.SetDirectParameters(ParameterView parameters)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewComponentFrame(DiffContext& diffContext, Int32 frameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewSubtree(DiffContext& diffContext, Int32 frameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InsertNewFrame(DiffContext& diffContext, Int32 newFrameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InsertNewFrame(DiffContext& diffContext, Int32 newFrameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.ComputeDiff(Renderer renderer, RenderBatchBuilder batchBuilder, Int32 componentId, ArrayRange`1 oldTree, ArrayRange`1 newTree)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.ProcessRenderQueue()
```

### Expected behavior

I would have expected that `width` would be treated as "unmatched attribute" and would go into `[Parameter( CaptureUnmatchedValues = true )] public Dictionary<string, object> Attributes { get; set; }`. But for some reason, it tries to convert it into a component parameter with the "same" name. Except for the uppercase.

I have looked through Blazor code and I think the problem might be in `SimplifiedStringHashComparer` on line: https://github.com/dotnet/aspnetcore/blob/79e5320951ab56b62ec18c06874ad7a1bd58555c/src/Components/Components/src/Rendering/SimplifiedStringHashComparer.cs#L23

Maybe I'm wrong and looking at the wrong place. In any case, I believe the casing should matter, and `width` and `Width` should be treated differently.

### To Reproduce

I have created a small project to reproduce the issue. It uses only the most basic types but the problem stays the same as in my main project.

Repo: https://github.com/stsrki/BlazorSensitiveParams

Look for the `Button` component, and test it on the`"/counter"` page.

### Further technical details

- ASP.NET Core version: 5.0
- VS 2019 Version 16.10.0 Preview 2.1

**Comment by stsrki** [comment]
Trying to find a workaround. It is ugly and I un-optimized but it seems to work

```cs
public override Task SetParametersAsync( ParameterView parameters )
{
    if ( parameters.TryGetValue<string>( "width", out var width ) )
    {
        var paremetersDictionary = parameters.ToDictionary() as Dictionary<string, object>;

        paremetersDictionary.Remove( "width" );

        Attributes = new Dictionary<string, object>();
        Attributes.Add( "width", width );

        return base.SetParametersAsync( ParameterView.FromDictionary( paremetersDictionary ) );
    }

    return base.SetParametersAsync( parameters );
}
```

It is far from the real solution. The same thing can happen basically for any other attribute.

**Comment by javiercn** [comment]
@stsrki thanks for contacting us.

This has been a core behavior from the framework since the beginning and not something we plan to change as that will potentially break all existing Blazor applications. We recommend you don't rely on casing for differentiating between parameters and instead just use different names.

We understand that this might not match your expectations, however there will be a cohort of people that will think they should have the current behavior and changing it now will only contribute to break existing applications and to upset a different set of people using Blazor.

If you want to have this behavior within your components we recommend you implement your own base class where you can decide the best way to map parameters to properties.

**Comment by stsrki** [comment]
I was afraid you would answer like that. But I think this is wrong. Changing the parameter names would lead to some weird naming decisions. Not just my own library but for any other out-there. The name `Width` is self-descriptive. Everyone who sees it knows what it means immediately. Sure, I could change it, but that is just hiding the real problem. As we all know C# language is case sensitive, and so should be the **razor** syntax.

Not sure if you remember, but we had a similar problem over two years ago when Blazor was still young. Back then components couldn't be named with an uppercase, eg `Button` component would be seen as a `button` element. This problem was fixed and you can see it in 3.0 preview 8 release notes: https://devblogs.microsoft.com/aspnet/asp-net-core-and-blazor-updates-in-net-core-3-0-preview-8/#case-sensitive-component-binding
